usr.bin/copong/copong.1
.\"
.\" Copyright (c) 2018 Edward Tomasz Napierala <en322@cl.cam.ac.uk>
.\" All rights reserved.
.\"
.\" This software was developed by SRI International and the University of
.\" Cambridge Computer Laboratory under DARPA/AFRL contract (FA8750-10-C-0237)
.\" ("CTSRD"), as part of the DARPA CRASH research programme.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd December 1, 2020
.Dt COPONG 1
.Os
.Sh NAME
.Nm copong
.Nd colocated ping responder
.Sh SYNOPSIS
.Nm
.Op Fl k
.Op Fl s Ar time
.Op Fl v
.Op Fl x
.Ar service-name
.Op Ar target-name ...
.Sh DESCRIPTION
The
.Nm
utility is used to answer to colocated pings from
.Xr coping 1
and
.Xr syscall_timing 1 .
If
.Ar target-name
is specified, upon receiving a ping it will cocall into each of the targets
before returning.
.Pp
The following options are available:
.Bl -tag -width ".Fl v"
.It Fl k
Use
.Xr coaccept_slow 2
instead of
.Xr coaccept 2 .
.It Fl s Ar time
Number of seconds to pause before calling
.Xr coaccept 2 .
Defaults to zero.
.It Fl v
Be verbose.
.It Fl x
Call
.Xr abort 3
at an unfortunate moment.
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh EXAMPLES
Execute in background within the address space of the current
shell process:
.Dl coexec Ar $$ Ar /usr/bin/copong target-name &
.Pp
.Sh SEE ALSO
.Xr coping 1 ,
.Xr coaccept 2 ,
.Xr colocation 7
.Sh HISTORY
The
.Nm
command appeared in
.Tn CheriBSD .
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was developed by
.An Edward Tomasz Napierala Aq Mt trasz@FreeBSD.org
at the University of Cambridge Computer Laboratory under DARPA/AFRL contract
(FA8750-10-C-0237) ("CTSRD"), as part of the DARPA CRASH research programme.


usr.bin/copong/copong.c
/*-
 * Copyright (c) 2018 Edward Tomasz Napierala <trasz@FreeBSD.org>
 * All rights reserved.
 *
 * This software was developed by SRI International and the University of
 * Cambridge Computer Laboratory under DARPA/AFRL contract (FA8750-10-C-0237)
 * ("CTSRD"), as part of the DARPA CRASH research programme.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/param.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void
usage(void)
{

	fprintf(stderr, "usage: copong [-s time] [-kxv] service-name [target-name ...]\n");
	exit(0);
}

int
main(int argc, char **argv)
{
	char inout[BUFSIZ];
	void * __capability cookie;
	void * __capability *lookedup;
	char *registered, *tmp;
	uint64_t *halfcookie;
	ssize_t received, outlen;
	pid_t pid;
	float dt = 0.0;
	bool kflag = false, vflag = false, xflag = false;
	int c, ch, error;

	while ((ch = getopt(argc, argv, "ks:xv")) != -1) {
		switch (ch) {
		case 'k':
			kflag = true;
			break;
		case 's':
			dt = strtof(optarg, &tmp);
			if (*tmp != '\0')
				errx(1, "argument to -s must be a number");
			if (dt < 0)
				errx(1, "argument to -s must be >= 0.0");
			break;
		case 'x':
			xflag = true;
			break;
		case 'v':
			vflag = true;
			break;
		case '?':
		default:
			usage();
		}
	}

	argc -= optind;
	argv += optind;
	if (argc < 1)
		usage();

	registered = argv[0];

	if (vflag)
		fprintf(stderr, "%s: %s: setting up...\n",
		    getprogname(), registered);
	error = cosetup(COSETUP_COACCEPT);
	if (error != 0)
		err(1, "cosetup");

	argc--;
	argv++;

	if (argc > 0) {
		if (vflag)
			fprintf(stderr, "%s: %s: setting up the caller side...\n",
			    getprogname(), registered);
		error = cosetup(COSETUP_COCALL);
		if (error != 0)
			err(1, "cosetup");

		lookedup = malloc(argc * sizeof(void * __capability));
		if (lookedup == NULL)
			err(1, "malloc");

		for (c = 0; c < argc; c++) {
			if (vflag)
				fprintf(stderr, "%s: %s: colooking up \"%s\"...\n",
				    getprogname(), registered, argv[c]);
			error = colookup(argv[c], &lookedup[c]);
			if (error != 0) {
				if (errno == ESRCH) {
					warnx("received ESRCH; this usually means there's nothing coregistered for \"%s\"", argv[c]);
					warnx("use coexec(1) to colocate; you might also find \"ps aux -o vmaddr\" useful");
				}
				err(1, "colookup");
			}
		}
	}

	if (vflag)
		fprintf(stderr, "%s: %s: coregistering...\n", getprogname(), registered);
	error = coregister(registered, NULL);
	if (error != 0)
		err(1, "coregister");

	if (vflag)
		fprintf(stderr, "%s: %s: coaccepting...\n", getprogname(), registered);

	memset(inout, 0, sizeof(*inout));
	received = 0; /* Nothing to send back at this point. */

	for (;;) {
		if (kflag)
			received = coaccept_slow(&cookie, inout, received, inout, sizeof(inout));
		else
			received = coaccept(&cookie, inout, received, inout, sizeof(inout));
		if (received < 0) {
			warn("%s", kflag ? "coaccept_slow" : "coaccept");
			received = 0;
			continue;
		}
		if (vflag) {
			halfcookie = (uint64_t *)&cookie;
			error = cocachedpid(&pid, cookie);
			if (error != 0)
				warn("cogetpid");
			printf("%s: %s: coaccepted, pid %d, caller pid %d, cookie %#lx%lx\n",
			    getprogname(), registered, getpid(), pid, halfcookie[0], halfcookie[1]);
		}

		/*
		 * XXX: Obviously only one of those responses will make it back to our caller.
		 */
		outlen = received;
		for (c = 0; c < argc; c++) {
			if (vflag)
				fprintf(stderr, "%s: %s: cocalling \"%s\"...\n",
				    getprogname(), registered, argv[c]);
			if (kflag)
				received = cocall_slow(lookedup[c], inout, outlen, inout, sizeof(inout));
			else
				received = cocall(lookedup[c], inout, outlen, inout, sizeof(inout));
			if (received < 0)
				warn("cocall");
			if (vflag)
				printf("%s: %s: returned from \"%s\", pid %d\n",
				    getprogname(), registered, argv[c], getpid());
		}

		if (dt > 0)
			usleep(dt * 1000000);

		if (xflag)
			abort();
	}
}
